### PR TITLE
fix(qwen3): pin decode attention-path to SplitKv under --batch-invariant

### LINKED ISSUE #613: qwen3: decode attention-path selection is bucket-driven, so batch invariance leaks under --batch-invariant
## Description

`BatchDecodeBuffers::attention_path(padded_bs)` picks the decode attention kernel purely from the padded batch bucket — `padded_bs <= 32` selects SplitKv, otherwise NonPartition — and never consults `NumericPolicy`. SplitKv and NonPartition are numerically different reductions (SplitKv merges partial softmax across KV chunks; NonPartition reduces in one pass per request), so a request that decodes at bucket <= 32 alone but is co-batched past 32 flips path, and its logits can move with batch composition. This survives `--batch-invariant` (Pin): Pin closes the projection-GEMM and split-KV chunk-count reduction-order axes, but the path *selection* itself is outside its reach.

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Fixes #613

`BatchDecodeBuffers::attention_path` picked the decode attention kernel from the padded batch bucket alone — `padded_bs <= SPLIT_KV_MAX_BATCH_SIZE` (32) selects SplitKv, above it NonPartition — without consulting `NumericPolicy`. SplitKv and NonPartition are numerically different reductions (SplitKv splits KV into fixed chunks and merges the online softmax; NonPartition reduces in one pass). So even under `--batch-invariant` (Pin), a request that decodes alone at a small bucket but is co-scheduled past bucket 32 switches kernel, and its logits move with its batch-mates.

## Fix

Under Pin and PerToken, `attention_path` returns SplitKv for every bucket instead of choosing by `padded_bs`. The split *size* is already request-local (#438 `pin_chunk_size`: chunk count = ceil(own_kv_len / fixed_size)), so pinning the path makes the whole attention step batch-invariant by construction — a request runs the same kernel with the same chunk count alone or co-batched. Tuned keeps the bucket-driven choice, byte-identical.

The split-KV workspace and CSR sizing follow so SplitKv is valid above bucket 32 under Pin — selection, workspace, metadata move together (leaving any one un-gated is a latent OOB / missing-CSR). Cost: +448 MiB device memory at MAX_BATCH=256 TP1 (split_tmp_v dominates), pre-subtracted from the KV cache under `--batch-invariant`; TP shrinks it (TP8 ~64 MiB). Pin-only; default Tuned pays nothing.

## Verification — Single GPU (x86_64, sm_89)

New gate `batch_invariance_attention_path`: request A prefilled isolated (batch 1), then decoded alone at bs=1 (SplitKv) vs co-batched at bs=40 (crosses bucket 32); assert A's decode top-K. Under Pin the GEMM-N (#428) and split-chunk (#438) axes are already fixed, so the only cross-batch drift left is the attention path — this isolates it.

A/B (base = attention_path bucket-driven, fix = this branch), same gate both arms:

| policy state | base Pin | fix Pin |
| --- | --- | --- |
| decode top-K eq (bs=1 vs bs=40) | false (drifts) | true (bit-identical) |

Tuned drifts in both arms with identical logprobs (base and fix baseline lp0 match exactly, so the Tuned path is unchanged). Regressions `batch_invariance_decode_gemm_graph` and `batch_invariance_endtoend` pass on fix.

Throughput A/B (`attention_path_perf`, Pin, decode tok/s, fix SplitKv>32 vs base NonPartition>32). It is a context-length-dependent trade-off, not a wash.

CTX=256 (short) — `net` = `delta` minus the bs=32 control bias (+2.7%, fix arm runs first):

| bs | fix | base | delta | net |
| --- | --- | --- | --- | --- |
| 32 (control) | 2547 | 2479 | +2.7% | — |
| 64 | 4063 | 3716 | +9.3% | +6.6% |
| 128 | 4294 | 3989 | +7.6% | +4.9% |
| 256 | 4150 | 4084 | +1.6% | -1.1% |

CTX=1024 (long) — control bias only +0.3% (clean), so the raw delta below is ~= net (no separate net column, unlike CTX=256's +2.7% bias):

| bs | fix | base | delta |
| --- | --- | --- | --- |
| 32 (control) | 1841 | 1835 | +0.3% |
| 40 | 1967 | 1980 | -0.7% |
| 64 | 2353 | 2445 | -3.7% |

At short context SplitKv wins (up to +6.6% net at bs=64), the gain narrowing to ~0 by MAX_BATCH. At long context SplitKv is throughput-negative above the cap (-3.7% at bs=64/CTX=1024): chunk count scales with context (1024/64 = 16 chunks/req), so merge overhead grows. This is a Pin opt-in trade-off (default Tuned is byte-identical).

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: openinfer-qwen3/tests/batch_invariance_attention_path.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+/// Prefill A alone (isolated KV), then decode it co-batched with `n_cofill` short fillers; return
+/// A's row-0 decode top-K. A's prefill is isolated, so only the decode batch size — and thus
+/// `attention_path` — changes across calls.
+fn a_decode_at_batch(ex: &mut Qwen3Executor, a_prompt: &[u32], n_cofill: usize) -> Vec<(u32, f32)> {
+    let id_a = RequestId::new(1);
+    let pr_a = ex
+        .execute_prefill(PrefillPlan {
+            sample_seed: 0,
+            requests: &[pitem(id_a, a_prompt.to_vec())],
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply openinfer-qwen3/tests/batch_invariance_attention_path.rs]: Great. Fixed.

File: openinfer-qwen3/tests/batch_invariance_attention_path.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+/// Whether A's decode top-K is identical at bs=1 vs the boundary-crossing batch, under `policy`.
+fn run_policy(policy: NumericPolicy, model_path: &str) -> bool {
+    set_numeric_policy(policy);
+    let mut ex = Qwen3Executor::from_runtime(model_path, true, &[0]).expect("build executor");
+    ex.set_prefix_cache_enabled(false);
+    let a = filler(A_LEN, 3);
+
+    let tk_small = a_decode_at_batch(&mut ex, &a, 0);
+    let tk_large = a_decode_at_batch(&mut ex, &a, LARGE_COFILL);
+    let eq = tk_small == tk_large;
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply openinfer-qwen3/tests/batch_invariance_attention_path.rs:144]: Fine. Done.
